### ISSUE TITLE
Support for elm 0.18

### ISSUE THREAD

**Issue by rosenk** [debois/elm-mdl#253]
Elm 0.18 have just been released. When whoud we expect a support for it?

**Comment by danstn** [comment]
Playing with migration to 0.18.

`elm-dom` and `elm-parts` left.

[Progress here](https://github.com/danstn/elm-mdl/commit/1009e713bd2c559cfbee4e892f72e067c2a2ba63)

**Comment by MichaelCombs28** [comment]
I got some time to get this done now, I'll start with elm-parts since we're using elm-mdl in production.

**Comment by danstn** [comment]
This is great news! Can't wait

**Comment by domenkozar** [comment]
@debois

**Comment by MichaelCombs28** [comment]
Already a pull request for elm-parts, just waiting on Debois

**Comment by danstn** [comment]
Yep, PRs for both `elm-dom` and `elm-parts`.

**Comment by domenkozar** [comment]
@MichaelCombs28 you can continue from https://github.com/danstn/elm-mdl/commit/1009e713bd2c559cfbee4e892f72e067c2a2ba63 and make a PR that uses the elm-dom and elm-parts fork. We'll help you test once elm-mdl compile.

**Comment by MichaelCombs28** [comment]
I'm gonna spend a bit of time writing a test suite for elm-parts, something weird happened when trying to compile it after migrating, specifically it didn't recognize the Index alias. Maybe there's a possibility that List ints are no longer comparables.

**Comment by MichaelCombs28** [comment]
https://github.com/MichaelCombs28/elm-mdl

working here, if anyone wants to help out, feel free!

**Comment by MichaelCombs28** [comment]
Finished the port and is ready for testing, working off https://github.com/MichaelCombs28/elm-mdl/tree/master
Maybe I should have made a new branch, but I wanted to finish fast so I forgot.

Also, please note that I'm using my forks as deps, but either way it's all done and should be ok to merge, if we don't get a response I'll publish mdl until Debois gets back to us.

**Comment by dpinn** [comment]
Michael, please do publish. I'd like to use your fork until it has been merged into debois/elm-mdl. Many thanks for your efforts.

**Comment by MichaelCombs28** [comment]
Done, just keep in mind I haven't ported the examples yet, the API is pretty much exactly the same, a few functions have changed & docs haven't been fully updated either but for the most part, any function named function' with the ' is now function_. http://package.elm-lang.org/packages/MichaelCombs28/elm-mdl/latest

**Comment by danstn** [comment]
Whoah! Great stuff :+1: Thanks heaps @MichaelCombs28  :facepunch:

**Comment by dpinn** [comment]
I don't want to be the bearer of bad news, but...

```
Problem in dependency MichaelCombs28/elm-mdl 1.0.0

The elm-package.json constraints of 'MichaelCombs28/elm-mdl' are probably
letting too much stuff through. Definitely open an issue on the relevant github
repo to get this fixed and save other people from this pain.

In the meantime, take a look through the direct dependencies of the broken
package and see if any of them have had releases recently. If you find the new
thing that is causing problems, you can artificially constrain things by adding
some extra constraints to your elm-package.json as a stopgap measure.
```

Does someone understand what that error message means?

**Comment by MichaelCombs28** [comment]
Strange I got it as well, it's only after you try to compile a project importing it. I'll deal with it in the morning.

**Comment by MichaelCombs28** [comment]
Just fixed it, guess I got a little obsessed, but yeah all done give it a try.

**Comment by dpinn** [comment]
Yes, that seems to have done the trick. No more error message. Many thanks, Michael.

**Comment by nsb** [comment]
Somehow I still get the error message for elm-parts.

    ERROR in ./src/elm/Main.elm
    Module build failed: Error: Compiler process exited with error Compilation failed
    Problem in dependency MichaelCombs28/elm-parts 1.0.0

    The elm-package.json constraints of 'MichaelCombs28/elm-parts' are probably
    letting too much stuff through. Definitely open an issue on the relevant github
    repo to get this fixed and save other people from this pain

**Comment by MichaelCombs28** [comment]
Can you show me your code? I'm not getting that error, even the examples are working now.

**Comment by nsb** [comment]
Yep, sorry. I found that I still need to update my code some more to work with 0.18. Somehow I didn't get those errors with my webpack setup only when manually running elm make. It's probably a problem  with my code.

**Comment by MichaelCombs28** [comment]
For now, any issues should be opened on my fork until we can merge. https://github.com/MichaelCombs28/elm-mdl/issues

Thanks for everyone's feedback!

**Comment by CIB** [comment]
As described [in this issue](https://github.com/MichaelCombs28/elm-mdl/issues/1), it's currently not possible to use the new elm 0.18 history serialization feature with elm-mdl.

**Comment by MarkNijhof** [comment]
@MichaelCombs28 I saw this message in the reply to @CIB 
> I'm in the planning stages of building a Material Design library for Elm using the sortable table concepts and style.
I was wondering how close you plan to keep the API? I mean I am about to start a project and want to use MDL but if it is going to be abandoned and an upgrade to the "proper" way is going to be very hard then I don't know what is wise to do. What would you recommend? My Elm knowledge atm is not at a level that would make contributing useful I am afraid. I want to port our React app using Material-ui to Elm and MDL seems a perfect fit.

**Comment by MichaelCombs28** [comment]
Sorry for not responding sooner, but the elm-mdl project doesn't belong to me, I just needed to upgrade it to 0.18 for a project. There's a good chance that if Debois doesn't come back to it we'll keep fixing things since it is being used in production right now.

The rework of it is my personal project, so to make this clear, there is a high level of certainty that we will keep working on it. Issues should be posted onto my fork until the merge. I am only following this thread so if you have an issue, please post it on my fork.

**Comment by MarkNijhof** [comment]
Thanks! That sounds promising. Do you have a timeline in mind for the port to 0.18 as I would really like to start on that vs 0.17. I noticed for the demo to run there needs to be an update to the router component being used. If anything new pops up I'll post to your fork.

**Comment by MichaelCombs28** [comment]
The port is already finished 

https://github.com/MichaelCombs28/elm-mdl

**Comment by MarkNijhof** [comment]
Yes, but not the demo application in there right? I am sorry if I wasn't
clear. I think having that would help for documentation and help. Besides
the router do you expect many changes in how to use MDL? I.e. How hard
would the upgrade of the demo app be?

I saw there is a pr for the router to put it on 0.18 if I can figure out
how to use that then I will make a stab at upgrading the demo app.

**Comment by MichaelCombs28** [comment]
The reason I haven't upgraded the Demo app is because of the changes to elm-navigation. The demo used a version of elm-route that relied on the old version of elm navigation. Other than that the api is pretty much the same, you must also account for the function that were used infix as before, I've changed them to account for composition in new Elm.

**Comment by debois** [comment]
Dear all, 

I've been away for too long, apologies. 

I'm in the process of upgrading to 0.18. The new version will feature (a) dropping the Parts library which seem to make so many people uncomfortable, and (b) multiple event dispatch courtesy of @vipentti. (I.e., if you want the focus event from a textfield, you can get it. Previously, you couldn't because the textfield needed it internally)

@MichaelCombs28: I tried to start off your fork, but you seem to have lost all @vipentti's work on multiple event dispatching, and it was very hard to merge with my draft no-parts variant, so much as I liked to, I couldn't start there. Appreciate you doing the fork!

@All: Apologies for the long absence.

**Comment by MarkNijhof** [comment]
Cool!

Do you have a timeline in mind? I am currently rebuilding via the MDC
css/js (also a great learning opportunity) as Google said they would not be
continuing development on MDL have you any thoughts on that?

-Mark

**Comment by MichaelCombs28** [comment]
Well if you need any help let me know! I've been on holiday for a few weeks and I'm back now as well so let us know the plan and we can help, this is possibly the most popular user created package outside of elm-community. Thank you so much for your efforts.

**Comment by debois** [comment]
@MichaelCombs28: Thanks, and you too! I'm good with the 0.18 release, I think, but in the short-term, elm-mdl really urgently need an autocomplete and a dropdown component; in the longer-term, we need to upgrade to Google MDL v2, which, if I'm understanding correctly, is actually the MDC that @MarkNijhof is referencing?

**Comment by hakonrossebo** [comment]
Would seem ok to close this issue now. Maybe create a new "roadmap" issue?

**Comment by cjduncana** [comment]
Sure, as long as the road map issue is referenced here.

**Comment by debois** [comment]
Makes sense.
